raft/src/main/java/org/apache/kafka/raft/KafkaRaftLog.java

```diff
@@ -0,0 +1,843 @@
+                log,
+                time,
+                scheduler,
+                recoverSnapshots(log),
+                topicPartition,
+                config,
+                nodeId
+        );
+
+        // Print a warning if users have overridden the internal config
+        if (config.logSegmentMinBytes() != KafkaRaftClient.MAX_BATCH_SIZE_BYTES) {
+            metadataLog.logger.error("Overriding " + MetadataLogConfig.METADATA_LOG_SEGMENT_MIN_BYTES_CONFIG +
+                    " is only supported for testing. Setting this value too low may lead to an inability to write batches of metadata records.");
+        }
+
+        // When recovering, truncate fully if the latest snapshot is after the log end offset. This can happen to a follower
+        // when the follower crashes after downloading a snapshot from the leader but before it could truncate the log fully.
+        metadataLog.truncateToLatestSnapshot();
+
+        return metadataLog;
+    }
+
+    public KafkaRaftLog(
+            UnifiedLog log,
+            Time time,
+            Scheduler scheduler,
+            // Access to this object needs to be synchronized because it is used by the snapshotting thread to notify the
+            // polling thread when snapshots are created. This object is also used to store any opened snapshot reader.
+            TreeMap<OffsetAndEpoch, Optional<FileRawSnapshotReader>> snapshots,
+            TopicPartition topicPartition,
+            MetadataLogConfig config,
+            int nodeId) {
+        this.log = log;
+        this.time = time;
+        this.scheduler = scheduler;
+        this.snapshots = snapshots;
+        this.topicPartition = topicPartition;
+        this.config = config;
+        this.logIdent = "[MetadataLog partition=" + topicPartition + ", nodeId=" + nodeId + "] ";
+        this.logger = new LogContext(logIdent).logger(KafkaRaftLog.class);
+    }
+
+    // for testing
+    UnifiedLog log() {
+        return log;
+    }
+
+    @Override
+    public LogFetchInfo read(long startOffset, Isolation readIsolation) {
+        FetchIsolation isolation = switch (readIsolation) {
+            case COMMITTED -> FetchIsolation.HIGH_WATERMARK;
+            case UNCOMMITTED -> FetchIsolation.LOG_END;
+        };
+
+        try {
+            FetchDataInfo fetchInfo = log.read(startOffset, config.maxFetchSizeInBytes(), isolation, true);
+            return new LogFetchInfo(
+                    fetchInfo.records,
+                    new LogOffsetMetadata(
+                            fetchInfo.fetchOffsetMetadata.messageOffset,
+                            Optional.of(new SegmentPosition(
+                                    fetchInfo.fetchOffsetMetadata.segmentBaseOffset,
+                                    fetchInfo.fetchOffsetMetadata.relativePositionInSegment))
+                    )
+            );
+        } catch (IOException ioe) {
+            throw new UncheckedIOException(ioe);
```

Comment:
Maybe it's a newcomer's question, is UncheckedIOException better than IOException?